flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java

```diff
@@ -98,115 +112,203 @@ private ScanContext(boolean caseSensitive, Long snapshotId, Long startSnapshotId
-  }
-
   Long splitOpenFileCost() {
     return splitOpenFileCost;
   }
 
-  ScanContext splitOpenFileCost(Long fileCost) {
-    return new ScanContext(caseSensitive, snapshotId, startSnapshotId, endSnapshotId, asOfTimestamp, splitSize,
-        splitLookback, fileCost, nameMapping, projectedSchema, filterExpressions);
+  boolean isStreaming() {
+    return isStreaming;
   }
 
-  String nameMapping() {
-    return nameMapping;
+  Duration monitorInterval() {
+    return monitorInterval;
   }
 
-  ScanContext nameMapping(String mapping) {
-    return new ScanContext(caseSensitive, snapshotId, startSnapshotId, endSnapshotId, asOfTimestamp, splitSize,
-        splitLookback, splitOpenFileCost, mapping, projectedSchema, filterExpressions);
+  String nameMapping() {
+    return nameMapping;
   }
 
   Schema projectedSchema() {
     return projectedSchema;
   }
 
-  ScanContext project(Schema schema) {
-    return new ScanContext(caseSensitive, snapshotId, startSnapshotId, endSnapshotId, asOfTimestamp, splitSize,
-        splitLookback, splitOpenFileCost, nameMapping, schema, filterExpressions);
-  }
-
   List<Expression> filterExpressions() {
     return filterExpressions;
   }
 
-  ScanContext filterRows(List<Expression> filters) {
-    return new ScanContext(caseSensitive, snapshotId, startSnapshotId, endSnapshotId, asOfTimestamp, splitSize,
-        splitLookback, splitOpenFileCost, nameMapping, projectedSchema, filters);
+  ScanContext copyWithAppendsBetween(long newStartSnapshotId, long newEndSnapshotId) {
+    return ScanContext.builder()
+        .caseSensitive(caseSensitive)
+        .useSnapshotId(null)
+        .startSnapshotId(newStartSnapshotId)
+        .endSnapshotId(newEndSnapshotId)
+        .asOfTimestamp(null)
+        .splitSize(splitSize)
+        .splitOpenFileCost(splitOpenFileCost)
+        .nameMapping(nameMapping)
+        .streaming(isStreaming)
+        .monitorInterval(monitorInterval)
+        .build();
+  }
+
+  ScanContext copyWithSnapshotId(long newSnapshotId) {
+    return ScanContext.builder()
+        .caseSensitive(caseSensitive)
+        .useSnapshotId(newSnapshotId)
+        .startSnapshotId(null)
+        .endSnapshotId(null)
+        .asOfTimestamp(null)
+        .splitSize(splitSize)
+        .splitOpenFileCost(splitOpenFileCost)
+        .nameMapping(nameMapping)
+        .streaming(isStreaming)
+        .monitorInterval(monitorInterval)
+        .build();
+  }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  static class Builder {
+    private boolean caseSensitive;
+    private Long snapshotId;
+    private Long startSnapshotId;
+    private Long endSnapshotId;
+    private Long asOfTimestamp;
+    private Long splitSize;
+    private Integer splitLookback;
+    private Long splitOpenFileCost;
+    private boolean isStreaming;
+    private Duration monitorInterval;
+    private String nameMapping;
+    private Schema projectedSchema;
+    private List<Expression> filterExpressions;
```

Comment:
Shouldn't a builder supply the defaults for these? Otherwise, all of the values need to be supplied. Why not use `CASE_SENSITIVE.defaultValue()` and similar here?